feat(requests): support caching response data in request result (#443)

oschlegel opened the pull request:
Introduce cacheResponseData to options for trackRequestResult function. This allows to cache the response data and return it in consecutive calls of the cached request

#443

## PR Checklist

Please check if your PR fulfills the following requirements:

- [/] The commit message follows our guidelines: https://github.com/ngneat/elf/blob/master/CONTRIBUTING.md#commit
- [/] Tests for the changes have been added (for bug fixes / features)
- [/] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

For skipped requests the trackRequestResult operator returns an empty observable.

Issue Number: 443

## What is the new behavior?

For skipped requests the trackRequestResult operator returns the same data (or error) like the original request. This behavior needs to be activated by using the cacheResponseData options of the trackRequestResult operator.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

NetanelBasal commented on packages/requests/src/lib/requests-result.spec.ts:
```diff
@@ -434,6 +438,174 @@ describe('requests result', () => {
   });
 
+  it('should cache response data', () => {
+    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));
```
Why do you need this line?

oschlegel replied:
I copy pasted form the "should work" test

oschlegel replied:
I guess it's to avoid the stale time

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -26,12 +28,13 @@ export interface LoadingRequestResult extends BaseRequestResult {
   isError: false;
   status: 'success';
   fetchStatus: 'idle';
+  data?: TData;
```
I don't like this name because it "conflicts" with the `data` from the `join` operator.

oschlegel replied:
renamed to responseData

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -198,13 +203,26 @@ export function trackRequestResult<TData>(
         ) {
-          return EMPTY;
+          return getRequestResult(key).pipe(
+            switchMap((it) => {
```
Why you named it `it`?

oschlegel replied:
result is already declared in the upper scope

NetanelBasal replied:
Yes, but what is `it`? We can name it `currentKeyResult`

oschlegel replied:
I named it requestResult

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -53,9 +56,9 @@ export interface IdleRequestResult extends BaseRequestResult {
 }
 
-export type RequestResult<TError = any> =
+export type RequestResult<TError = any, TData = any> =
```
Add a test for the type, please

hakimio replied:
`expectTypeOf(value).toEqualTypeOf<Type>();`

NetanelBasal replied:
Yes ^^

NetanelBasal commented on packages/requests/src/lib/requests-result.spec.ts:
```diff
@@ -434,6 +438,174 @@ describe('requests result', () => {
+
+    const { state, config } = createState(withEntities<Todo>());
+
+    const store = new Store({ state, config, name: 'todos' });
```
We don't need all of these cases in this test, as we test only the cache property functionality. Keep it slim, please.

oschlegel replied:
Removed the addEntities instruction

NetanelBasal commented on packages/requests/src/lib/requests-result.spec.ts:
```diff
@@ -426,6 +434,152 @@ describe('requests result', () => {
+
+    const store = new Store({ state, config, name: 'todos' });
+
+    const entities$ = store.pipe(
```
Doing this isn't make sense for what you did. The whole purpose of this feature is NOT to use the joinRequestResult operator.

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -159,7 +162,7 @@ export function resetStaleTime(key: unknown[]) {
 export function joinRequestResult<T, TError = any>(
   ...[key, options]: Parameters<typeof getRequestResult>
-): OperatorFunction<T, RequestResult<TError> & { data: T }> {
+): OperatorFunction<T, RequestResult<TError, T> & { data: T }> {
```
This change doesn't makes sense. This `data` isn't the response data.

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -179,6 +182,8 @@ interface Options {
   skipCache?: boolean;
   // Skip the request if it's already fetching
   preventConcurrentRequest?: boolean;
+  // Cache the response data
```
```suggestion
  // Wheteher to cache the response data
```

oschlegel replied:
Adapted

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -198,13 +203,28 @@ export function trackRequestResult<TData>(
           !stale
         ) {
-          return EMPTY;
+          return getRequestResult(key).pipe(
```
There is no need to do all of that if cacheResponseData is false.

```
if(cacheResponseData) { ... } else { return EMPTY }
```

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -198,13 +203,28 @@ export function trackRequestResult<TData>(
-          return EMPTY;
+          return getRequestResult(key).pipe(
+            switchMap((requestResult) => {
+              if (requestResult.isError) {
```
What's the use case for isError? You'll get here only if it's success

oschlegel replied:
If `preventConcurrentRequest` is truthy this code can be reached for requests which are `fetching`, those requests can succeed or error

NetanelBasal replied:
So we need this chunk only in this case

NetanelBasal commented on packages/requests/src/lib/requests-result.spec.ts:
```diff
@@ -434,6 +443,96 @@ describe('requests result', () => {
+
+    const store = new Store({ state, config, name: 'todos' });
+
+    const entities$ = store.pipe(selectAllEntities());
```
Why do you need a store? You said that you want the data directly from the result

oschlegel replied:
I'm checking in the test if entites in the store still behave as expected when cacheResponseData is set

NetanelBasal commented on packages/requests/src/lib/requests-result.ts:
```diff
@@ -26,12 +28,13 @@ export interface LoadingRequestResult extends BaseRequestResult {
   isError: false;
   status: 'success';
   fetchStatus: 'idle';
+  responseData?: TData;
```
Note that when reading this value it will always be TData or undefined even if guard is use

oschlegel commented:
Maybe the trackRequestResult shoudn't return a empty obserable at all and instead return `responseData`, even when not cached. This way both version behave more similar

oschlegel commented:
@NetanelBasal do you still see doings here?

NetanelBasal commented:
Sorry, I'm very busy these days. Ping me next week, please.

oschlegel commented:
@NetanelBasal ping :)

oschlegel commented:
I added tests for the type. I changed trackRequestResult so that it returns the cached response (or undefined) instead of empty observable. @NetanelBasal could you have a look if this can now be merged?

NetanelBasal commented on docs/docs/features/requests-result.mdx:
```diff
@@ -126,6 +126,21 @@ export function fetchTodos() {
+import { trackRequestResult } from '@ngneat/elf-requests';
+import { setTodos } from './todos.repository';
+
+export function fetchTodos() {
```
Subscribe to fetchTodos and show that we can access the data property

oschlegel replied:
done

NetanelBasal commented on docs/docs/features/requests-result.mdx:
````diff
@@ -126,6 +126,21 @@ export function fetchTodos() {
 }
 ```
 
+- `cacheResponseData` - Cache the response data and emit it as data for skipped requests, defaults to `false`
````
```suggestion
- `cacheResponseData` - Cache the response data and emit it as responseData for skipped requests, defaults to `false`
```

oschlegel replied:
done
